From: Michael Tokarev <mjt@tls.msk.ru>
Date: Thu, 19 May 2022 20:37:21 +0300
Updated: Thu, 28 Nov 2024 11:59:58 +0300
Subject: specify some deps on private libs actually used directly
Forwarded: https://gitlab.com/samba-team/samba/-/merge_requests/3872

Lots of samba libraries has incomplete dependencies listed
in wscript files.  This usually is not a problem since the
link line includes dependencies of their dependencies of
their dependencies, and somewhere down that line all immediate
dependencies which are missing are actually present.  But
sometimes this becomes a problem when a library does not
declare direct dependency on at least one private library
which it actually uses: in case no private library is
listed as direct dependency, private library directory is
not put into RUNPATH of the resulting binary, so the binary
can not find its own dependencies.

Fix a few such places, including some libraries which are
a part of public abi (libsmbldap, libndr).

Signed-off-by: Michael Tokarev <mjt@tls.msk.ru>
---
 lib/param/wscript_build                             | 2 +-
 lib/util/wscript_build                              | 4 ++--
 librpc/wscript_build                                | 1 +
 source3/wscript_build                               | 4 ++--
 source4/dsdb/samdb/ldb_modules/wscript_build_server | 4 ++--
 source4/param/wscript_build                         | 2 +-
 6 files changed, 9 insertions(+), 8 deletions(-)

diff --git a/lib/param/wscript_build b/lib/param/wscript_build
index 864975a5884..0ea874b9f6e 100644
--- a/lib/param/wscript_build
+++ b/lib/param/wscript_build
@@ -34,3 +34,3 @@ bld.SAMBA_LIBRARY('server-role',
                   source='loadparm_server_role.c',
-                  deps='samba-util samba-debug',
+                  deps='samba-util time-basic samba-debug',
                   private_library=True)
diff --git a/lib/util/wscript_build b/lib/util/wscript_build
index b4fcfeaba07..38614702930 100644
--- a/lib/util/wscript_build
+++ b/lib/util/wscript_build
@@ -245,3 +245,3 @@ else:
                       source='modules.c',
-                      deps='samba-errors samba-util',
+                      deps='samba-errors samba-util samba-debug',
                       local_include=False,
@@ -279,3 +279,3 @@ else:
                       local_include=False,
-                      public_deps='tevent samba-errors',
+                      public_deps='tevent time-basic samba-errors',
                       public_headers='tevent_ntstatus.h tevent_unix.h tevent_werror.h',
diff --git a/librpc/wscript_build b/librpc/wscript_build
index 6cb4e1cff2b..e8e36118adb 100644
--- a/librpc/wscript_build
+++ b/librpc/wscript_build
@@ -674,2 +674,3 @@ bld.SAMBA_LIBRARY('ndr',
     public_deps='samba-errors talloc samba-util util_str_hex',
+    deps='genrand',
     public_headers='gen_ndr/misc.h gen_ndr/ndr_misc.h ndr/libndr.h:ndr.h',
diff --git a/source3/wscript_build b/source3/wscript_build
index 0c9f6ad3143..3ed3b550960 100644
--- a/source3/wscript_build
+++ b/source3/wscript_build
@@ -194,3 +194,3 @@ bld.SAMBA3_LIBRARY('smbldaphelper',
                           ''',
-                   deps='smbldap secrets3',
+                   deps='smbldap secrets3 samba-security samba3-util',
                    allow_undefined_symbols=True,
@@ -502,3 +502,3 @@ bld.SAMBA3_LIBRARY('smbldap',
                     source='lib/smbldap.c',
-                    deps='ldap lber samba-util smbconf',
+                    deps='ldap lber samba-util smbconf samba-security genrand smbd_shim',
                     enabled=bld.CONFIG_SET("HAVE_LDAP"),
diff --git a/source4/dsdb/samdb/ldb_modules/wscript_build_server b/source4/dsdb/samdb/ldb_modules/wscript_build_server
index 9c1eb12a7c2..1f70641283a 100644
--- a/source4/dsdb/samdb/ldb_modules/wscript_build_server
+++ b/source4/dsdb/samdb/ldb_modules/wscript_build_server
@@ -347,3 +347,3 @@ bld.SAMBA_MODULE('ldb_anr',
 	internal_module=False,
-	deps='talloc samba-util samdb'
+	deps='talloc samba-util samdb ldbsamba'
 	)
@@ -387,3 +387,3 @@ bld.SAMBA_MODULE('ldb_resolve_oids',
 	internal_module=False,
-	deps='samdb talloc ndr'
+	deps='samdb talloc ndr ldbsamba'
 	)
diff --git a/source4/param/wscript_build b/source4/param/wscript_build
index d1b852c0c73..80341369337 100644
--- a/source4/param/wscript_build
+++ b/source4/param/wscript_build
@@ -41,3 +41,3 @@ bld.SAMBA_PYTHON('pyparam',
 	         source='pyparam.c',
-	         deps='samba-hostconfig %s' % pytalloc_util,
+	         deps='samba-hostconfig server-role samba-util samba-debug %s' % pytalloc_util,
 	         realname='samba/param.so'
-- 
2.39.5

